### PR TITLE
Make each test class loaded by separate class loader (TestNG)

### LINKED ISSUE #608: Powermockito and testng breaks jacoco report on  few tests
Here is an example project: https://github.com/astafev/powermock-testng-demo
ClassA and ClassB. ClassATest patches ClassB, ClassBTest patches ClassA. As a result in runtime we have both class patched. If run only one test, normal report will be generated; if both - no coverage at all in the report.

It can be fixed if in org.powermock.modules.testng.internal.PowerMockClassloaderObjectFactory all classes will be created with new classloader.

### PR DESCRIPTION
**Pull request by astafev**
tested on https://github.com/astafev/powermock-testng-demo and real project

### CODE REVIEW

**Comment by johanhaleby** [comment]
Thanks a lot for trying to tackle this problem. I just have a question, won't this severely slow down the test execution??

**Comment by astafev** [comment]
I don't have big project on testNG to test, but with small project I don't see any difference (tests take about 40 seconds with both versions).

I think it depends on project - in current version we work with manipulated classes in all tests, as I understand, manipulated classes are slower. In new version only needed by particular test case classes are manipulated, other test cases aren't affected.

**Comment by astafev** [comment]
Hello, our project became bigger a bit, I still don't see difference in time of test run (66148ms with 1.6.4 vs 64448ms with patched). But I had to increase PermGen size in 2 times.

I wanted to ask: do you have some other doubts about the PR? Because I don't see here any new performance problems, except extra PermGen usage.

Found a new problem that this change can fix: if we run in parallel few tests and they use mockStatic, or just static variables, it can cause problems.
Here is a project that reproduces the problem: https://github.com/astafev/testng-methods-order

File: modules/module-impl/testng/src/main/java/org/powermock/modules/testng/PowerMockObjectFactory.java [hunk 1]
```diff
@@ -64,6 +66,9 @@ private boolean anyMethodInClassHasPowerMockAnnotation(Class<?> testClass) {
     }
 
     private boolean isClassAnnotatedWithPowerMockAnnotation(Class<?> testClass) {
-        return testClass.isAnnotationPresent(PrepareForTest.class) || testClass.isAnnotationPresent(SuppressStaticInitializationFor.class);
+        return testClass.isAnnotationPresent(PrepareForTest.class)
```
Comment (thekingnothing) [inline modules/module-impl/testng/src/main/java/org/powermock/modules/testng/PowerMockObjectFactory.java]: Please, cover by tests.

**Comment by thekingnothing** [comment]
1. I don't see a new test in the pull request. You've added the several annotation to `isClassAnnotatedWithPowerMockAnnotation` but didn't add test to cover it. 
2. If some new feature is being added, like run each class in separate class loader (By the way, I agree that it's sometime is needed), then first of all we need to think about how it could affect other users and do they really need it always. As for me, a develop should has ability to enable/disable this behaviour. Also if theatre is added to TestN module, same should be implemented for jUnit module. 
3. Have you tested how `@BeforeClass` and `@AfterClass` works after changes? Especial, in case if some static fields are initialised in these methods?

**Comment by astafev** [comment]
Hi, I'll split it into 2 pull requests. Found another place where @PrepareEverythingForTest should be added.

About junit - looks, like in junit it's already used. 
https://github.com/jayway/powermock/blob/0320072de8626fa984b6c60b895d12ea0e2625bb/modules/module-impl/junit4-rule/src/main/java/org/powermock/modules/junit4/rule/PowerMockClassloaderExecutor.java#L41

**Pull request merged**